eth/consensus/clique/_utils.py

```diff
@@ -0,0 +1,154 @@
+import time
+from typing import (
+    Iterable,
+)
+
+from eth.abc import (
+    BlockHeaderAPI,
+)
+from eth.constants import (
+    EMPTY_UNCLE_HASH,
+    ZERO_ADDRESS,
+    ZERO_HASH32,
+)
+from eth_keys import keys
+from eth_keys.datatypes import PrivateKey
+from eth_typing import (
+    Address,
+    Hash32,
+)
+from eth_utils import (
+    encode_hex,
+    to_list,
+    ValidationError,
+)
+
+from .constants import (
+    COMMON_ADDRESS_LENGTH,
+    NONCE_AUTH,
+    NONCE_DROP,
+    SIGNATURE_LENGTH,
+    VANITY_LENGTH,
+)
+from .datatypes import (
+    Snapshot,
+)
+
+
+@to_list
+def get_signers_at_checkpoint(header: BlockHeaderAPI) -> Iterable[Address]:
+    """
+    Read the list of signers from a checkpoint header.
+    """
+
+    signer_count = int(
+        (len(header.extra_data) - VANITY_LENGTH - SIGNATURE_LENGTH) / COMMON_ADDRESS_LENGTH
+    )
+
+    for i in range(signer_count):
+        yield Address(
+            header.extra_data[VANITY_LENGTH + i * COMMON_ADDRESS_LENGTH:][:COMMON_ADDRESS_LENGTH]
+        )
+
+
+def get_signature_hash(header: BlockHeaderAPI) -> Hash32:
+    """
+    Return the hash that is signed by the block producer. It is defined as the hash of
+    the ``header`` except that the last 65 bytes of the ``extra_data`` (the signature)
+    are removed before calculating the hash.
+    """
+    signature_header = header.copy(
+        extra_data=header.extra_data[:len(header.extra_data) - SIGNATURE_LENGTH]
+    )
+    return signature_header.hash
+
+
+def get_block_signer(header: BlockHeaderAPI) -> Address:
+    """
+    Return the address of the signer of the ``header``.
+    """
+
+    if len(header.extra_data) < SIGNATURE_LENGTH:
+        raise Exception("header.extra_data too short to contain signature")
```

Comment:
Seems like this check should also be in place within `get_signature_hash`, otherwise if the `extra_data` is too short, the signature hash will just be the header with empty `extra_data`.

Reply:
Sounds correct, in fact, we can just *move* the check there and then reorder the lines here so that `signature_hash = get_signature_hash(header)` comes first.